### ISSUE TITLE
Warning 6133 ignores @link tags

### ISSUE THREAD

**Issue by DanielSWolf** [microsoft/TypeScript#47558]
# Bug Report

### 🔎 Search Terms

* "@link"
* is declared but its value is never read. (6133)

### 🕗 Version & Regression Information

- This changed between versions 4.2.3 and 4.3. The problem  concerns the `@link` tag, which was introduced in version 4.3.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.5&multiFile=true#code/PTAEAEDMEsBsFMB2BDAtvAXKA7gC3vLAHQAuAzgFAUigAqu0Zos0i8ouyTJ+OyATolYBzUACIA5AGEBE0I1AATeAGNYA+ItAAjAK4l55UADdksXewVtj8fqH7xkiogAoAbAEYAzF4CUY6jB4G0RQHgB7XWFcAwADGX5Y+SYHSFskFU0dAE8w3nAWRABrMORhIgpoVAAHcP4DEmzq9gBvUATQAF9QSH5w1FAJImAVWQBuKmAAKinQAEEcfEIeutBdMnZsaB5QZFAWgtYShM6iUCngCngAD1r6+UQSW0hkTNAAdSXYfYpQUBoiICKJ1JmAoHAkGhMKBRvxSJQrrc6gZWE9+C83h0Wr9-mBARUQUA)

### 💻 Code

```ts
// @filename: wheel.ts

import type { Car } from './car';

/** A wheel for use with a {@link Car}. */
export interface Wheel {
  // ...
}

// @filename: car.ts

export interface Car {
  // ...
}
```

### 🙁 Actual behavior

The line `import type { Car } from './car';` has the following compiler warning:

> 'Car' is declared but its value is never read. (6133)

### 🙂 Expected behavior

There should be no compiler warning for the type import. The purpose of this warning is to indicate superfluous imports, and this type import isn't superfluous: It is required for the `@link` tag to work property.

Without the type import, TypeScript could not resolve the target of the link. As a result, the tooltip in VS code would contain the `Car` reference in plain text rather than as a link.

Compare the following screenshots:

*With type import:*
![](https://i.stack.imgur.com/6UPgc.png)

*Without type import:*
![](https://i.stack.imgur.com/nAAy5.png)

**Comment by RyanCavanaugh** [comment]
I'm not sure if this is going to be fixable per se - @sandersn do we parse link tags at all during TS checking?

**Comment by sandersn** [comment]
Link tags are parsed and are in principle bindable and checkable. Neither one happens today, though -- name resolution happens on a services-only code path.

Looking at the code that leads to "value is never read" errors, it looks like checking link tags would fix this. It would also put errors on unresolved links in Typescript, which I don't *think* is desirable. But (1) it might be, actually or (2) it would be pretty easy to squelch the error for link tags.

**Comment by DanielSWolf** [comment]
> Link tags are parsed and are in principle bindable and checkable.

That's great! 😀

> It would also put errors on unresolved links in Typescript, which I don't _think_ is desirable. But (1) it might be, actually or (2) it would be pretty easy to squelch the error for link tags.

From my point of view, it would certainly be desirable. The reason why I use the `@link` syntax is because I want TypeScript to create an actual link. If TypeScript can't do that because I'm missing an import, I find that important to know.

I regularly spend time double-checking generated documentation to make sure everything worked, and I certainly don't enjoy it. If TypeScript can support me by flagging unresolved links, I'd certainly appreciate it!

**Comment by JaegerJens** [comment]
Without TypeScript support writing public API documentation requires a lot of house keeping:

- Detecting dead links
- Rename refactoring breaks API documentation
- Workarounds for false TypeScript errors

**Comment by komyg** [comment]
I will try to fix this along with the other related issue that I am working on: [#47753](https://github.com/microsoft/TypeScript/issues/47753).

Here is the PR: https://github.com/microsoft/TypeScript/pull/47824

**Comment by willster277** [comment]
Had this pop up in my code today, 

Specifically what I want is for the emitted `.d.ts` files to maintain the imports, I don't much care about the JS files as I have `removeComments` enabled for them anyway.

I'm completely unfamiliar with TypeScript's code, but it's probably worth `removeComments = true` causing the compiler to ignore checking syntax within comments (for efficiency and to prevent accidentally flagging good code as invalid or vice versa), however this should only be at compile time. For example, using `@link` for internal documentation while having `removeComments = true` should ignore the `@link` when compiling, but respect it while checking the file in an editor such as VSCode.

Either way, I hope this can be resolved soon as I'm halfway to bodging a syntax & post processor to re-insert the imports at compile time.

**Comment by willster277** [comment]
@sandersn the error message in editor was;
`'imported_object_name' is declared but its value is never read. ts(6133)`
and in the compiler logs;
`file.ts: - error TS6133: 'imported_object_name' is declared but its value is never read.`

I believe the compiler error is from `"noUnusedLocals": true` as once it's imported, a type/object counts as a local variable for the file. Disabling the option removes the compiler error (but doesn't fix the actual problem), but the editor error still persists.

[Here is the repo reproducing the problem](https://github.com/willster277/ts-issue-link-tags), though looking through #47822 I believe it will be resolved. I haven't opened a new issue as it would simply duplicate this one.
Hopefully we'll be able to test if it's fixed on the next `@next` release!
